### PR TITLE
Save tag in task

### PR DESCRIPTION
**Pull request by t19106**

### CODE REVIEW

File: app/models/task.rb [hunk 1]
```diff
@@ -28,6 +30,18 @@ def less_than_a_day
     end
   end
 
+  def tasks_not_overlapping
+    return if starts_at.nil? || ends_at.nil?
+    tasks = Task.where(user: user_id, starts_at: self.starts_at.beginning_of_day..self.starts_at.end_of_day).or(Task.where(user: user_id, ends_at: self.ends_at.beginning_of_day..self.ends_at.end_of_day)).map { |task| task.adjust_overnight_range(self.starts_at) }
```
Comment (koheisg) [inline app/models/task.rb]: at first, extract method.
and name this process and try refactoring.

File: app/models/task.rb [hunk 2]
```diff
@@ -28,6 +30,18 @@ def less_than_a_day
...
+    tasks.each do |task|
+      if (task.starts_at..task.ends_at).cover?(self.starts_at) || (task.starts_at..task.ends_at).cover?(self.ends_at) || (self.starts_at..self.ends_at).cover?(task.starts_at) || (self.starts_at..self.ends_at).cover?(task.ends_at)
```
Comment (koheisg) [inline app/models/task.rb]: at first, extract method.
and name this process and try refactoring.

File: app/controllers/tasks/months_controller.rb [hunk 3]
```diff
@@ -2,8 +2,16 @@ class Tasks::MonthsController < ApplicationController
   before_action :verify_user, :validate_uri
 
   def index
-    tasks_by_start = Task.where(user: current_user, starts_at: @date.beginning_of_month..@date.end_of_month).map { |task| task.adjust_overnight_range(task.starts_at) }.group_by { |t| t.starts_at.day }
-    tasks_by_end = Task.where(user: current_user, ends_at: @date.beginning_of_month..@date.end_of_month).map { |task| task.adjust_overnight_range(task.ends_at) }.group_by { |t| t.ends_at.day }
     @tasks = tasks_by_start.merge(tasks_by_end) { |_k, v1, v2| v1.concat(v2).uniq.sort_by { |t| t.starts_at } }
```
Comment (koheisg) [inline app/controllers/tasks/months_controller.rb]: more readable by extract class.

File: app/controllers/tasks/months_controller.rb [hunk 4]
```diff
@@ -1,9 +1,21 @@
...
   def index
-    tasks_by_start = Task.where(user: current_user, starts_at: @date.beginning_of_month..@date.end_of_month).map { |task| task.adjust_overnight_range(task.starts_at) }.group_by { |t| t.starts_at.day }
-    tasks_by_end = Task.where(user: current_user, ends_at: @date.beginning_of_month..@date.end_of_month).map { |task| task.adjust_overnight_range(task.ends_at) }.group_by { |t| t.ends_at.day }
-    @tasks = tasks_by_start.merge(tasks_by_end) { |_k, v1, v2| v1.concat(v2).uniq.sort_by { |t| t.starts_at } }
   end
+
+  private
+
+    def tasks_by_start
+      Task.where(user: current_user, starts_at: @date.beginning_of_month..@date.end_of_month).map { |task| task.adjust_overnight_range(task.starts_at) }.group_by { |t| t.starts_at.day }
```
Comment (koheisg) [inline app/controllers/tasks/months_controller.rb]: business logic is to be written in model.

File: app/controllers/tasks/months_controller.rb [hunk 5]
```diff
@@ -1,9 +1,21 @@
...
+    end
+
+    def tasks_by_end
+      Task.where(user: current_user, ends_at: @date.beginning_of_month..@date.end_of_month).map { |task| task.adjust_overnight_range(task.ends_at) }.group_by { |t| t.ends_at.day }
```
Comment (koheisg) [inline app/controllers/tasks/months_controller.rb]: business logic is to be written in model.

File: app/controllers/tasks/months_controller.rb [hunk 6]
```diff
@@ -1,9 +1,21 @@
...
+    end
+
+    def set_monthly_tasks
+      @tasks = tasks_by_start.merge(tasks_by_end) { |_k, v1, v2| v1.concat(v2).uniq.sort_by { |t| t.starts_at } }
```
Comment (koheisg) [inline app/controllers/tasks/months_controller.rb]: business logic is to be written in model.

File: app/controllers/tasks/months_controller.rb [hunk 7]
```diff
@@ -1,9 +1,21 @@
 class Tasks::MonthsController < ApplicationController
   before_action :verify_user, :validate_uri
+  before_action :set_monthly_tasks, only: %i(index)
```
Comment (koheisg) [inline app/controllers/tasks/months_controller.rb]: `before_action` is before action. it has too much process.
only request parameter tuning.

File: app/models/task.rb [hunk 8]
```diff
@@ -45,14 +58,27 @@ def over_end_of_month?
...
     elsif ends_at > date.end_of_day
       _task.ends_at = ends_at.change(day: date.tomorrow.day, hour: 0, min: 0)
     end
     _task
   end
+
+  private
+
+    def same_day_tasks
+      Task.where(user: user_id, starts_at: self.starts_at.beginning_of_day..self.starts_at.end_of_day).or(Task.where(user: user_id, ends_at: self.ends_at.beginning_of_day..self.ends_at.end_of_day)).map { |task| task.adjust_overnight_range(self.starts_at) }
```
Comment (koheisg) [inline app/models/task.rb]: extract method.

File: app/models/task.rb [hunk 9]
```diff
@@ -45,14 +58,27 @@ def over_end_of_month?
...
+    end
+
+    def overlapping?(task)
+      (task.starts_at..task.ends_at).cover?(self.starts_at) || (task.starts_at..task.ends_at).cover?(self.ends_at) || (self.starts_at..self.ends_at).cover?(task.starts_at) || (self.starts_at..self.ends_at).cover?(task.ends_at)
```
Comment (koheisg) [inline app/models/task.rb]: extract method.

File: app/controllers/tasks/days_controller.rb [hunk 10]
```diff
@@ -2,7 +2,7 @@ class Tasks::DaysController < ApplicationController
   before_action :verify_user, :validate_uri
 
   def index
-    @tasks = Task.where(user: current_user, starts_at: @date.beginning_of_day..@date.end_of_day).or(Task.where(user: current_user, ends_at: @date.beginning_of_day..@date.end_of_day)).map { |task| task.adjust_overnight_range(@date) }.sort_by { |t| t.starts_at }
-    @task = Task.new
+    @tasks = Task.tasks_by_date(@date, current_user)
```
Comment (koheisg) [inline app/controllers/tasks/days_controller.rb]: where is `@date` declared?

File: app/models/task.rb [hunk 11]
```diff
@@ -28,31 +30,83 @@ def less_than_a_day
...
   end
 
-  def to_minutes
-    (((ends_at - starts_at) / 60) % 60).to_i
+  def self.tasks_by_month(date, user)
+    monthly_tasks_by_starts_at(date, user).merge(monthly_tasks_by_ends_at(date, user)) { |_k, v1, v2| v1.concat(v2).uniq.sort_by { |t| t.starts_at } }
+  end
+
+  def self.monthly_tasks_by_starts_at(date, user)
+    Task.where(user: user, starts_at: date.beginning_of_month..date.end_of_month).map { |task| task.adjust_overnight_range(task.starts_at) }.group_by { |t| t.starts_at.day }
```
Comment (koheisg) [inline app/models/task.rb]: I prefer `user.tasks.where(....`

File: app/models/task.rb [hunk 12]
```diff
@@ -28,31 +30,83 @@ def less_than_a_day
...
+  end
+
+  def to_minutes
+    (((ends_at - starts_at) / 1.minute.to_i) % 1.minute.to_i).to_i
   end
 
   def adjust_overnight_range(date)
-    return self unless is_overnight?
+    return self unless overnight?
     _task = self.clone
```
Comment (koheisg) [inline app/models/task.rb]: http://www.techscore.com/blog/2012/12/13/rails%E3%81%A7%E5%A4%89%E6%9B%B4%E5%89%8D%E3%81%AE%E5%80%A4%E3%80%81%E5%9E%8B%E5%A4%89%E6%8F%9B%E5%89%8D%E3%81%AE%E5%80%A4%E3%80%81%E3%81%AA%E3%81%A9%E3%81%AA%E3%81%A9/

File: app/models/task.rb [hunk 13]
```diff
@@ -28,31 +30,83 @@ def less_than_a_day
...
     if starts_at < date.beginning_of_day
-      _task.starts_at = starts_at.change(day: date.day, hour: 0, min: 0)
-      _task.starts_at = starts_at.change(month: date.month, day: date.day, hour: 0, min: 0) if over_end_of_month?
+      _task.starts_at = over_end_of_month? ? starts_at.change(month: date.month, day: date.day, hour: 0, min: 0) : starts_at.change(day: date.day, hour: 0, min: 0)
```
Comment (koheisg) [inline app/models/task.rb]: https://qiita.com/AKB428/items/20e81ccc8d9998b5535d#三項演算子

File: app/controllers/tasks/days_controller.rb (hunk 10 above)
Comment (fi1446) [reply app/controllers/tasks/days_controller.rb]: application_controller#validate_url

File: app/models/task.rb (hunk 12 above)
Comment (fi1446) [reply app/models/task.rb]: これは初耳です、活用してみます

**Comment by koheisg** [comment]
How about using the following as a reference for the extract method?

> Keep each line of code to a readable length. Unless you have a reason to, keep lines to fewer than 100 characters.
> https://github.com/github/rubocop-github/blob/master/STYLEGUIDE.md

File: app/models/task.rb (hunk 12 above)
Comment (fi1446) [reply app/models/task.rb]: よく読んでなかった。ようするに `self.clone` が今となっては無駄ってことね

File: app/models/task.rb [hunk 14]
```diff
@@ -1,58 +1,128 @@
...
     unless to_hours < 24
       errors.add(:ends_at, '24時間を超えるタスクは設定できません。')
     end
   end
 
-  def to_hours
-    ((ends_at - starts_at) / 60 / 60).to_i
+  def tasks_not_overlapping
+    return unless time_set?
+    user = User.find(user_id)
```
Comment (koheisg) [inline app/models/task.rb]: defined `belongs_to :user`, not need to `User.find`.

File: app/models/task.rb [hunk 15]
```diff
@@ -1,58 +1,128 @@
...
+  has_one :tags
 
   validates :starts_at, presence: true
   validates :ends_at, presence: true
   validate :ended_after_started
   validate :started_and_ended_are_different
   validate :less_than_a_day
+  validate :tasks_not_overlapping
 
   def ended_after_started
```
Comment (koheisg) [inline app/models/task.rb]: private

File: app/models/task.rb [hunk 16]
```diff
@@ -1,58 +1,128 @@
...
-    return if starts_at.nil? || ends_at.nil?
+    return unless time_set?
     if starts_at > ends_at
       errors.add(:ends_at, '終了時間は開始時間より前に設定できません。')
     end
   end
 
   def started_and_ended_are_different
```
Comment (koheisg) [inline app/models/task.rb]: private

File: app/models/task.rb [hunk 17]
```diff
@@ -1,58 +1,128 @@
...
-    return if starts_at.nil? || ends_at.nil?
+    return unless time_set?
     if starts_at == ends_at
       errors.add(:ends_at, '開始時間と終了時間が同じです。0分以上の作業時間を登録しましょう。')
     end
   end
 
   def less_than_a_day
```
Comment (koheisg) [inline app/models/task.rb]: private

File: app/models/task.rb [hunk 18]
```diff
@@ -1,58 +1,128 @@
...
-    return if starts_at.nil? || ends_at.nil?
+    return unless time_set?
     unless to_hours < 24
       errors.add(:ends_at, '24時間を超えるタスクは設定できません。')
     end
   end
 
-  def to_hours
-    ((ends_at - starts_at) / 60 / 60).to_i
+  def tasks_not_overlapping
```
Comment (koheisg) [inline app/models/task.rb]: private

File: app/models/task.rb [hunk 19]
```diff
@@ -1,58 +1,128 @@
...
+    user.tasks.where(starts_at: date.beginning_of_day..date.end_of_day)
   end
 
-  def over_end_of_month?
-    starts_at.month != ends_at.month
+  def self.same_day_tasks_by_ends_at(date, user)
+    user.tasks.where(ends_at: date.beginning_of_day..date.end_of_day)
+  end
+
+  def to_hours
```
Comment (koheisg) [inline app/models/task.rb]: private

File: app/models/task.rb [hunk 20]
```diff
@@ -1,58 +1,128 @@
...
+    ((ends_at - starts_at) / 1.hour.to_i).to_i
+  end
+
+  def to_minutes
```
Comment (koheisg) [inline app/models/task.rb]: private

File: app/controllers/concerns/parameter_datetime.rb [hunk 21]
```diff
@@ -0,0 +1,16 @@
...
+    begin
+      Time.zone.local(params[:year], params[:month], params[:day])
+    rescue ArgumentError
+      flash[:alert] = '存在しない日付へアクセスされようとしました。今月のページへ戻ります。'
+      redirect_to tasks_months_path year: Time.zone.now.year, month: Time.zone.now.month
+    end
+  end
+
+  def params_datetime
+    @date = Time.zone.local(params[:year], params[:month], params[:day])
```
Comment (koheisg) [inline app/controllers/concerns/parameter_datetime.rb]: `@date` への代入って必要？

File: app/models/task.rb [hunk 22]
```diff
@@ -1,58 +1,128 @@
...
+          errors.add(:ends_at, '他のタスクと時間が重複しています。')
+        end
+      end
+    end
   end
 
-  def to_minutes
-    (((ends_at - starts_at) / 60) % 60).to_i
+  def self.tasks_by_month(date, user)
+    monthly_tasks_by_starts_at(date, user).merge(monthly_tasks_by_ends_at(date, user)) { |_k, v1, v2| v1.concat(v2).uniq.sort_by { |t| t.starts_at } }
```
Comment (koheisg) [inline app/models/task.rb]: too much length

File: app/models/task.rb [hunk 23]
```diff
@@ -1,58 +1,128 @@
...
+  end
+
+  def self.monthly_tasks_by_starts_at(date, user)
+    user.tasks.where(starts_at: date.beginning_of_month..date.end_of_month).map { |task| task.adjust_overnight_range(task.starts_at) }.group_by { |t| t.starts_at.day }
```
Comment (koheisg) [inline app/models/task.rb]: too much length

File: app/models/task.rb [hunk 24]
```diff
@@ -1,58 +1,128 @@
...
+  end
+
+  def self.monthly_tasks_by_ends_at(date, user)
+    user.tasks.where(ends_at: date.beginning_of_month..date.end_of_month).map { |task| task.adjust_overnight_range(task.ends_at) }.group_by { |t| t.ends_at.day }
```
Comment (koheisg) [inline app/models/task.rb]: too much length

File: app/models/task.rb [hunk 25]
```diff
@@ -1,58 +1,128 @@
...
+  end
+
+  def self.tasks_by_date(date, user)
+    same_day_tasks_by_starts_at(date, user).or(same_day_tasks_by_ends_at(date, user)).map { |task| task.adjust_overnight_range(task.starts_at) }.sort_by { |t| t.starts_at }
```
Comment (koheisg) [inline app/models/task.rb]: too much length

File: app/models/task.rb (hunk 20 above)
Comment (fi1446) [reply app/models/task.rb]: to_hoursとto_minutesは外部で使います

File: app/controllers/concerns/parameter_datetime.rb (hunk 21 above)
Comment (fi1446) [reply app/controllers/concerns/parameter_datetime.rb]: Viewで使うから必要、とは言いたいけど、それはこのモジュールの責任でないということかな

**Pull request merged**